kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java

```diff
@@ -680,4 +650,99 @@ public CloudConnectionStatusEnum getNotificationStatus() {
         this.notificationStatus = status;
     }
+
+    private final class PublishManager implements Runnable {
+
+        @Override
+        public void run() {
+            Thread.currentThread().setName("DataServiceImpl:Submit");
+            while (publisherEnabled) {
+                long sleepingTime = -1;
+                if (DataServiceImpl.this.dataTransportService.isConnected()) {
+                    try {
+                        DataMessage message = DataServiceImpl.this.store.getNextMessage();
+
+                        if (message != null) {
+                            checkInFlightMessages(message);
+
+                            boolean messagePublished;
+                            if (dataServiceOptions.isRateLimitEnabled() && message.getPriority() >= 5) {
+                                messagePublished = publishMessageTokenBucket(message);
+                            } else {
+                                messagePublished = publishMessageUnbound(message);
+                            }
+
+                            if (messagePublished) {
+                                continue;
+                            } else {
+                                sleepingTime = DataServiceImpl.this.throttle.getNextRefillEta();
+                            }
+                        }
+                    } catch (KuraNotConnectedException e) {
+                        logger.info("DataPublisherService is not connected");
+                    } catch (KuraTooManyInflightMessagesException e) {
+                        logger.info("Too many in-flight messages");
+                        handleInFlightCongestion();
+                    } catch (Exception e) {
+                        logger.error("Probably an unrecoverable exception", e);
+                    }
+                } else {
+                    logger.info("DataPublisherService not connected");
+                }
+
+                try {
+                    suspendPublisher(sleepingTime, TimeUnit.MILLISECONDS);
+                } catch (InterruptedException e) {
+                    logger.debug("Interrupting publishing thread.");
+                    Thread.currentThread().interrupt();
+                }
+            }
+            logger.debug("Exited publisher loop.");
+        }
+
+        private void checkInFlightMessages(DataMessage message) throws KuraTooManyInflightMessagesException {
+            if (message.getQos() > 0
+                    && DataServiceImpl.this.inFlightMsgIds.size() >= dataServiceOptions.getMaxInFlightMessages()) {
+                logger.warn("The configured maximum number of in-flight messages has been reached");
+                throw new KuraTooManyInflightMessagesException("Too many in-flight messages");
+            }
+        }
+
+        private void suspendPublisher(long timeout, TimeUnit timeUnit) throws InterruptedException {
+            try {
+                lock.lock();
+                if (!publisherEnabled) {
```

Comment:
Why ``publisherEnabled`` must be checked in a lock?
You don't need to throw an Exception, just return.